### DISCUSSION TITLE
Using an old USB keyboard with Meshtastic node.

### DISCUSSION THREAD

**Discussion by so1der** [meshtastic/Discussions#111] [General]
Hello there! So I managed to use old membrane keyboard to type and send messages via Meshtastic node.
You can see details [in project repo](https://github.com/so1der/mesh_usb_keyboard).

Those USB membrane keyboard can also work as PS/2 keyboards. PS/2 protocol is fair simple, so almost any microcontroller can work with it. Unfortunately wireless keyboards (even with USB dongle) wont work, as well as the mechanical ones I suppose.

This method imitates cardkb, so we dont need to modify Meshtastic firmware, but we do need additional microcontroller. I used Arduino Nano. Since Arduino and keyboard utilize 5V, we also need Logic Level Shifter.

Maybe its possible to implement PS/2 protocol into Meshtastic, so we can at least get rid of additional microcontroller?

![schematic](https://github.com/user-attachments/assets/2fe8d0b9-f95f-4245-938e-695475df33dd)

**Comment by easytarget** [comment]
I really like the thinking behind this; althought it is specific for very old mechanical keyboards the basic technique is very sound; leverage the existing support for the I2C cardKB. to interface to other keyboards.. 

Most keyboards you would want to use for meshtastic are going to be more modern and smaller; and based on the USB and bluetooth HID (Human Interface Device) standards.. But until meshtastic itself directly supports HID this is a great workaround.

You could probably swap out the 5v Nano for a 3.3v part with support for USB host modes; the ATmega32U4 based Arduino boards can do this (and are well priced).  RP2040/50, and later ESP32 models such as the S2 and S3 also have on-chip support for USB host modes. 

I imagine the Code from the OP could be adaped to swap out the PS2 driver stuff for a HID device and translator. With ESP based boards you can possibly also use Bluetooth HID interfaces.

**Comment by mofosyne** [comment]
Did some quick study of existing code for PS/2 hosts libaries that looks most relevant, one link that came up is  https://github.com/techpaul/PS2KeyAdvanced as well as https://github.com/PaulStoffregen/PS2Keyboard

I think it's not a bad idea to have an option of plugging in a USB keyboard as a quick field input option. In the field, for the foreseeable future there will still be lots of PCs with a normal hard wired keyboard attached. However it does mean for practicality purpose your 'keyboard input' socket should have a Type-C and a Type-A port unfortunately... but that depends on how prepared you want your device to be.

Plus having to boost a 3.3v to 5v is annoying...

**Reply by easytarget** [reply]
What really attracts me is the idea of using a bluetooth keyboard.

The esp idf provides bluetooth HID services:
https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/bluetooth/esp_hidh.html
And
https://github.com/espressif/esp-idf/tree/v5.3.2/examples/bluetooth/esp_hid_host

It should be possible to embed a small bt->i2c keyboard adapter in a case and then carry a cheap-ass bluetooth travel keyboard around. A C3 or S3 esp chip could do this and also be acceptably low power if programmed properly.

I'd love to work on this myself.. It would really suit my use-case, but I'm waaaay overloaded as-is :-(

**Reply by mofosyne** [reply]
@easytarget can't we directly connect Bluetooth mestastic to Bluetooth keyboard directly?

**Reply by garthvh** [reply]
> @easytarget can't we directly connect Bluetooth mestastic to Bluetooth keyboard directly?

No most Bluetooth keyboards are classic Bluetooth, meshtastic only supports BLE.

**Reply by easytarget** [reply]
> most Bluetooth keyboards are classic Bluetooth

Most portable Bluetooth keyboards, the sort sold for use with mobile phones and tablets are BLE5/5.1 It is *this* sort of keyboard that meshtastic could support; they are ideal for carrying around with a portable node. They are also ubiquitous; I have [one of these](https://www.action.com/nl-nl/p/3203225/lab31-opvouwbaar-draadloos-toetsenbord/).. note the price.

But..while I think it iis technically possible, that does not mean it should be done...
... It's oh-so easy to 'request' this, but someone would need to code it, test it, squeeze it into an already crowded code-space, maintain it, etc.. So expect it to stay on the 'if we had a million dollars' wish-list, although it would be a good fit with the UI development work. I think the core developers have much more important things on their hands right now.

Final thought...
If you have a BT keyboard and a phone with the meshtastic app on it;  _you dont need the keyboard connected to the node, it is better to have it connected to the phone.._ The app is much nicer than the node itself for composing and browsing messages.

**Comment by raxpww** [comment]
connected PS/2 keyboard directly to ESP32 to work via network LoRa https://www.youtube.com/watch?v=lM5eNa6sBSg
